base/dates/io.jl

```diff
@@ -261,8 +265,99 @@ Characters not listed above are normally treated as delimiters between date and
 For example a `dt` string of "1996-01-15T00:00:00.0" would have a `format` string like
 "y-m-dTH:M:S.s". If you need to use a code character as a delimiter you can escape it using
 backslash. The date "1995y01m" would have the format "y\\ym\\m".
+
+Creating a DateFormat object is expensive. Whenever possible, create it once and use it many times
+or try the `dateformat""` string macro. Using this macro creates the DateFormat object once at
+macro expansion time and reuses it later. see [`@dateformat_str`](@ref).
+
+See [`DateTime`](@ref) and [`format`](@ref) for how to use a DateFormat object to parse and write Date strings
+respectively.
 """
-DateTime(dt::AbstractString,format::AbstractString;locale::AbstractString="english") = DateTime(dt,DateFormat(format,locale))
+function DateFormat(f::AbstractString, locale::DateLocale=ENGLISH)
+    tokens = AbstractDateToken[]
+    prev = ()
+    prev_offset = 1
+
+    letters = String(collect(keys(Base.Dates.SLOT_RULE)))
+    for m in eachmatch(Regex("(?<!\\\\)([\\Q$letters\\E])\\1*"), f)
+        tran = replace(f[prev_offset:m.offset-1], r"\\(.)", s"\1")
+
+        if !isempty(prev)
+            letter, width = prev
+            typ = SLOT_RULE[letter]
+
+            push!(tokens, DatePart{letter}(width, isempty(tran)))
+        end
+
+        if !isempty(tran)
+            push!(tokens, Delim(length(tran) == 1 ? first(tran) : tran))
+        end
+
+        letter = f[m.offset]
+        width = length(m.match)
+
+        prev = (letter, width)
+        prev_offset = m.offset + width
+    end
+
+    tran = replace(f[prev_offset:endof(f)], r"\\(.)", s"\1")
+
+    if !isempty(prev)
+        letter, width = prev
+        typ = SLOT_RULE[letter]
+
+        push!(tokens, DatePart{letter}(width, false))
+    end
+
+    if !isempty(tran)
+        push!(tokens, Delim(length(tran) == 1 ? first(tran) : tran))
+    end
+
+    tokens_tuple = (tokens...)
+    return DateFormat{Symbol(f),typeof(tokens_tuple)}(tokens_tuple, locale)
+end
+
+function DateFormat(f::AbstractString, locale::AbstractString)
+    DateFormat(f, LOCALES[locale])
+end
+
+function Base.show(io::IO, df::DateFormat)
+    write(io, "dateformat\"")
+    for t in df.tokens
+        _show_content(io, t)
+    end
+    write(io, '"')
+end
+
+"""
+    dateformat"Y-m-d H:M:S"
+
+Create a [`DateFormat`](@ref) object. Similar to `DateFormat("Y-m-d H:M:S")`
+but creates the DateFormat object once during macro expansion.
+
+See [`DateFormat`](@ref) for details about format specifiers.
+"""
+macro dateformat_str(str)
+    DateFormat(str)
+end
+
+# Standard formats
+const ISODateTimeFormat = DateFormat("yyyy-mm-dd\\THH:MM:SS.s")
+const ISODateFormat = DateFormat("yyyy-mm-dd")
+const RFC1123Format = DateFormat("e, dd u yyyy HH:MM:SS")
```

Comment:
So it works by looking at `Delim{'Z'}`? When someone is not using TimeZones, Base parses till before Z and everything keeps working? Kinda feels like a hack, but if it works... Feel free to push this change if you see it makes sense.